### PR TITLE
Work on NEON intrinsics

### PR DESCRIPTION
**Pull request by math-gout**
This PR aims to refactor the NEON part of `machine_vectors.h`, optimize some functions and implement `vec2n_addmod`, `vec2n_addmod_limited` and `vec2n_bit_shift_right`.

The `vec2n_bit_shift_right` part is very messy and I don't really like it. Let me know if I should revert it.

### CODE REVIEW

File: src/machine_vectors.h [hunk 1]
```diff
@@ -1370,134 +1381,212 @@ DEFINE_IT(vec2d)
...
+FLINT_FORCE_INLINE vec1n vec1d_convert_limited_vec1n(vec1d a) {
+    return (slong)a;
 }
 
-FLINT_FORCE_INLINE void vec4n_store_unaligned(ulong* z, vec4n a) {
-    vec2n_store_unaligned(z+0, a.e1);
-    vec2n_store_unaligned(z+2, a.e2);
+// (a + b) % n
+FLINT_FORCE_INLINE vec1n vec1n_addmod(vec1n a, vec1n b, vec1n n) {
+    double nmb = n - b;
```
Comment (fredrik-johansson) [inline src/machine_vectors.h:1396]: Should be ``ulong``, no?

Comment (math-gout) [reply src/machine_vectors.h:1396]: Yes ! Good catch !

**Comment by fredrik-johansson** [comment]
This needs to be checked (for correctness and performance) by someone with access to an arm64 machine which I currently don't.

Test code for the stuff in ``machine_vectors.h`` would clearly be useful to have...

The ``shift_right`` code looks like more trouble than it's worth since we currently only use this operation with a constant 32-bit shift. If we ever need it with a variable shift, the table lookup is going to hurt anyway.

**Comment by math-gout** [comment]
> This needs to be checked (for correctness and performance) by someone with access to an arm64 machine which I currently don't.

I have an M1 MacBook Air, a Raspberry Pi 4 and an Odroid C2 which are all arm64 I could test this on.
How should I go about testing the performance ?

> The shift_right code looks like more trouble

Completely agree, I will remove it.

**Comment by math-gout** [comment]
All tests pass on the M1 and the Raspberry Pi 4. As for the Odroid C2, I sadly do not have access to it right now.

**Comment by math-gout** [comment]
Haven't touched this for a while, so I tried to rebase, but it looks like it didn't go very well ...
I think it will be easier to close this and start fresh.

**Comment by fredrik-johansson** [comment]
Yeah, there is some issue with rebasing in a PR. I don't know how to avoid that.

**Comment by math-gout** [comment]
I managed to apply my changes on top of the main branch, now every thing seem to work.

**Comment by albinahlback** [comment]
Have you done any benchmarks?

**Comment by math-gout** [comment]
No I haven't. What would you suggest running ?

**Comment by albinahlback** [comment]
I think I created `src/fft_small/profile/p-fft_small_vs_gmp.c` to estimate the cutoff point between using GMP and our `fft_small` module. If you can show that this PR lowers this cutoff or is about the same point, that would be perfect.

You may have to play with the preprocessor constants defined in that file. You can compile all profiler files for `fft_small` via `make profile MOD=fft_small`, and then run said profiler via `./build/fft_small/profile/p-fft_small_vs_gmp`.

**Comment by math-gout** [comment]
Thank's !
I had to add `for (int i = 0; i < 1000; i ++)` before each `mul` function to get better precision because otherwise I got mainly `inf` or `nan`.
Here is what I got with the main branch :
```
mpn_mul_n vs fft_small
n = 1500: 1.66
n = 1505: 1.39
n = 1510: 1.43
n = 1515: 1.44
n = 1520: 1.42
n = 1525: 1.42
n = 1530: 1.44
n = 1535: 1.43
n = 1540: 1.37
n = 1545: 1.37
n = 1550: 1.36
n = 1555: 1.36
n = 1560: 1.37
n = 1565: 1.37
n = 1570: 1.42
n = 1575: 1.36
n = 1580: 1.37
n = 1585: 1.38
n = 1590: 1.38
n = 1595: 1.36
n = 1600: 1.32

mpn_sqr vs fft_small
n = 2900: 1.36
n = 2910: 1.36
n = 2920: 1.35
n = 2930: 1.34
n = 2940: 1.33
n = 2950: 1.42
n = 2960: 1.47
n = 2970: 1.47
n = 2980: 1.46
n = 2990: 1.45
n = 3000: 1.44
n = 3010: 1.40
n = 3020: 1.45
n = 3030: 1.44
n = 3040: 1.42
n = 3050: 1.44
n = 3060: 1.42
n = 3070: 1.42
n = 3080: 1.54
n = 3090: 1.54
n = 3100: 1.59
n = 3110: 1.59
n = 3120: 1.57
n = 3130: 1.61
n = 3140: 1.58
n = 3150: 1.57
n = 3160: 1.57
n = 3170: 1.58
n = 3180: 1.52
n = 3190: 1.57
n = 3200: 1.57
```

And here is what I got with this PR :
```
mpn_mul_n vs fft_small
n = 1500: 1.68
n = 1505: 1.38
n = 1510: 1.48
n = 1515: 1.44
n = 1520: 1.42
n = 1525: 1.42
n = 1530: 1.41
n = 1535: 1.43
n = 1540: 1.39
n = 1545: 1.40
n = 1550: 1.41
n = 1555: 1.36
n = 1560: 1.35
n = 1565: 1.42
n = 1570: 1.37
n = 1575: 1.41
n = 1580: 1.40
n = 1585: 1.42
n = 1590: 1.40
n = 1595: 1.40
n = 1600: 1.33

mpn_sqr vs fft_small
n = 2900: 1.37
n = 2910: 1.36
n = 2920: 1.35
n = 2930: 1.36
n = 2940: 1.36
n = 2950: 1.43
n = 2960: 1.49
n = 2970: 1.45
n = 2980: 1.46
n = 2990: 1.46
n = 3000: 1.45
n = 3010: 1.43
n = 3020: 1.44
n = 3030: 1.44
n = 3040: 1.44
n = 3050: 1.42
n = 3060: 1.43
n = 3070: 1.40
n = 3080: 1.57
n = 3090: 1.59
n = 3100: 1.60
n = 3110: 1.60
n = 3120: 1.59
n = 3130: 1.64
n = 3140: 1.61
n = 3150: 1.59
n = 3160: 1.60
n = 3170: 1.57
n = 3180: 1.55
n = 3190: 1.60
n = 3200: 1.58
```

On main `mul` starts to win at `n = 720` and `sqr` at `n = 800`
On this PR `mul` starts to win at `n = 715` and `sqr` at `n = 790`

**Comment by albinahlback** [comment]
Ops, sorry for the long reply. Thanks for this!

**Pull request merged**